File: app/src/test/java/com/codemate/brewflop/DayUpdateSchedulerTest.java
package com.codemate.brewflop;

import android.app.AlarmManager;
import android.content.Context;
import android.content.Intent;

import com.codemate.brewflop.alarms.AlarmReceiver;
import com.codemate.brewflop.alarms.DayUpdateScheduler;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.shadows.ShadowAlarmManager;
import org.robolectric.shadows.ShadowPendingIntent;

import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class DayUpdateSchedulerTest {
    private Context context;
    private ShadowAlarmManager shadowAlarmManager;

    private DayUpdateScheduler dayUpdateScheduler;

    @Before
    public void setUp() {
        context = RuntimeEnvironment.application.getApplicationContext();

        AlarmManager manager = (AlarmManager) RuntimeEnvironment.application.getSystemService(Context.ALARM_SERVICE);
        shadowAlarmManager = shadowOf(manager);

        dayUpdateScheduler = new DayUpdateScheduler(context);
    }

    @Test
    public void shouldScheduleAlarmEveryMorningAtSix() {
        assertNull(shadowAlarmManager.getNextScheduledAlarm());

        dayUpdateScheduler.setAlarm();

        ShadowAlarmManager.ScheduledAlarm scheduledAlarm = shadowAlarmManager.getNextScheduledAlarm();
        assertNotNull(scheduledAlarm);

        assertThat(scheduledAlarm.interval, is(Constants.DAY_UPDATE_INTERVAL));
        assertThat(scheduledAlarm.triggerAtTime, is(dayUpdateScheduler.getTriggerTime() + Constants.DAY_UPDATE_INTERVAL));
        assertThat(scheduledAlarm.type, is(AlarmManager.RTC_WAKEUP));
    }

    @Test
    public void shouldOnlyHaveOneAlarmAtTime() {
        dayUpdateScheduler.setAlarm();
        dayUpdateScheduler.setAlarm();
        dayUpdateScheduler.setAlarm();

        assertThat(shadowAlarmManager.getScheduledAlarms().size(), is(1));
    }

    @Test
    public void shouldCallAlarmReceiverWhenTimeIsUp() {
        Intent expectedIntent = new Intent(context, AlarmReceiver.class);

        dayUpdateScheduler.setAlarm();

        ShadowAlarmManager.ScheduledAlarm scheduledAlarm = shadowAlarmManager.getNextScheduledAlarm();
        ShadowPendingIntent shadowPendingIntent = shadowOf(scheduledAlarm.operation);

        assertTrue(shadowPendingIntent.isBroadcastIntent());
        assertThat(shadowPendingIntent.getSavedIntents().length, is(1));
        assertThat(shadowPendingIntent.getSavedIntents()[0].getComponent(), is(expectedIntent.getComponent()));
    }
}

File: app/src/test/java/com/codemate/brewflop/alarms/DayUpdateSchedulerTest.kt
package com.codemate.brewflop.alarms

import android.app.AlarmManager
import android.content.Context
import android.content.Intent
import com.codemate.brewflop.Constants
import junit.framework.Assert.*
import org.hamcrest.CoreMatchers
import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.Shadows.shadowOf
import org.robolectric.shadows.ShadowAlarmManager

@RunWith(RobolectricTestRunner::class)
class DayUpdateSchedulerTest {
    private lateinit var context: Context
    private lateinit var shadowAlarmManager: ShadowAlarmManager
    private lateinit var dayUpdateScheduler: DayUpdateScheduler

    @Before
    fun setUp() {
        context = RuntimeEnvironment.application.applicationContext

        val manager = RuntimeEnvironment.application.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        shadowAlarmManager = shadowOf(manager)

        dayUpdateScheduler = DayUpdateScheduler(context)
    }

    @Test
    fun shouldScheduleAlarmEveryMorningAtSix() {
        assertNull(shadowAlarmManager.nextScheduledAlarm)

        dayUpdateScheduler.setAlarm()

        val scheduledAlarm = shadowAlarmManager.nextScheduledAlarm
        assertNotNull(scheduledAlarm)

        assertThat(scheduledAlarm.interval, CoreMatchers.`is`(Constants.DAY_UPDATE_INTERVAL))
        assertThat(scheduledAlarm.triggerAtTime, `is`(dayUpdateScheduler.triggerTime + Constants.DAY_UPDATE_INTERVAL))
        assertThat(scheduledAlarm.type, `is`(AlarmManager.RTC_WAKEUP))
    }

    @Test
    fun shouldOnlyHaveOneAlarmAtTime() {
        dayUpdateScheduler.setAlarm()
        dayUpdateScheduler.setAlarm()
        dayUpdateScheduler.setAlarm()

        assertThat(shadowAlarmManager.scheduledAlarms.size, `is`(1))
    }

    @Test
    fun shouldCallAlarmReceiverWhenTimeIsUp() {
        val expectedIntent = Intent(context, AlarmReceiver::class.java)

        dayUpdateScheduler.setAlarm()

        val scheduledAlarm = shadowAlarmManager.nextScheduledAlarm
        val shadowPendingIntent = shadowOf(scheduledAlarm.operation)

        assertTrue(shadowPendingIntent.isBroadcastIntent)
        assertThat(shadowPendingIntent.savedIntents.size, `is`(1))
        assertThat(shadowPendingIntent.savedIntents[0].component, `is`(expectedIntent.component))
    }
}